File: 백준/Silver/15655. N과 M （6）/N과 M （6）.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const [N, M] = input[0].split(" ").map(Number);
const nums = input[1]
  .split(" ")
  .map(Number)
  .sort((a, b) => a - b);

const sequence = [];

const dfs = (count, start) => {
  if (count === M) {
    console.log(sequence.join(" "));
    return;
  }

  for (let i = start; i < N; i++) {
    sequence.push(nums[i]);
    start++;
    dfs(count + 1, start);
    sequence.pop();
  }
};

dfs(0, 0);

File: 백준/Silver/15655. N과 M （6）/N과 M （6）.py
N, M = map(int, input().split())
nums = list(map(int, input().split()))
stack = []
visited = [0] * (N+1)

nums.sort()


def dfs(depth):
    if depth == M:
        print(*stack)
        return

    for idx, num in enumerate(nums):
        if stack and stack[-1] >= num:
            continue
        if not visited[idx]:
            visited[idx] = 1
            stack.append(num)
            dfs(depth+1)
            stack.pop()
            visited[idx] = 0


dfs(0)
